src/Components/Order/class-shop-order-service.php

```diff
@@ -0,0 +1,273 @@
+		$order->setOrderNumber( $wc_order->get_order_number() );
+		$order->setTotalPrice( $wc_order->get_total() );
+		$order->setShippingPrice( $wc_order->get_shipping_total() );
+		$order->setItems( $this->get_order_items( $wc_order ) );
+
+		$order->setBillingAddress( $this->get_billing_address( $wc_order ) );
+		$order->setShippingAddress( $this->get_shipping_address( $wc_order ) );
+		$shipping_method = Shipping_Method_Helper::get_packlink_shipping_method_from_order( $wc_order );
+		if ( null !== $shipping_method ) {
+			$order->setShippingMethodId( $shipping_method->getId() );
+		}
+
+		$drop_off_point_id = $this->get_drop_off_point_id( (int) $order_id );
+		if ( $drop_off_point_id ) {
+			$order->setShippingDropOffId( $drop_off_point_id );
+		}
+
+		return $order;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function updateTrackingInfo( $order_id, Shipment $shipment, array $tracking_history ) {
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function updateShipmentStatus( $order_id, $shipping_status ) {
+		$order      = $this->get_order_by_id( $order_id );
+		$status_map = $this->configuration->getOrderStatusMappings();
+		$old_status = $order->get_status();
+		if ( isset( $status_map[ $shipping_status ] ) && $status_map[ $shipping_status ] !== $old_status ) {
+			$order->set_status( $status_map[ $shipping_status ], __( 'Status set by Packlink PRO.', 'packlink-pro-shipping' ) );
+		}
+
+		$order->save();
+	}
+
+	/**
+	 * Returns order instance, if exists.
+	 *
+	 * @param string $order_id $orderId Unique order id.
+	 *
+	 * @return WC_Order WooCommerce order object.
+	 * @throws OrderNotFound When order with provided id is not found.
+	 */
+	public function get_order_by_id( $order_id ) {
+		$wc_order = \WC_Order_Factory::get_order( $order_id );
+		if ( false === $wc_order ) {
+			/* translators: %s: order identifier */
+			throw new OrderNotFound( sprintf( __( 'Order with id(%s) not found!', 'packlink-pro-shipping' ), $order_id ) );
+		}
+
+		return $wc_order;
+	}
+
+	/**
+	 * Returns category name.
+	 *
+	 * @param \WC_Product $product WooCommerce product.
+	 *
+	 * @return string|null Category name.
+	 */
+	private function get_product_category_name( \WC_Product $product ) {
+		$category_ids = $product->get_category_ids();
+		if ( empty( $category_ids ) ) {
+			return null;
+		}
+
+		$category = WP_Term::get_instance( $category_ids[0] );
+
+		return $category instanceof WP_Term ? $category->name : null;
+	}
+
+	/**
+	 * Returns array of formatted order items.
+	 *
+	 * @param WC_Order $wc_order WooCommerce order.
+	 *
+	 * @return Item[] Array of formatted order items.
+	 */
+	private function get_order_items( WC_Order $wc_order ) {
+		$items = array();
+		/**
+		 * WooCommerce order item.
+		 *
+		 * @var \WC_Order_Item_Product $wc_item
+		 */
+		foreach ( $wc_order->get_items() as $wc_item ) {
+			$product = $wc_item->get_product();
+			if ( $product->is_downloadable() || $product->is_virtual() ) {
```

Comment:
Did you consider to use array_filter? If you don't need all array items probably foreach is not the best way to solve it

Reply:
This is exactly the `array_filter` functionality but without one unnecessary loop over all array items.

Reply:
You could combine array_filter and array_map it improve readability because you would be separating two responsibilities in two functions . Probably performance between this solution and implement this two responsibilities in foreach is the same.

Reply:
Not in my opinion. When one looks at this code, one can clearly see the condition for this code to be executed. Introducing more functions and logic than a simple `if` decreases the readability, is more error-prone and could add more problems than it is ought to solve.

Reply:
In this code we have two separate responsibilities (Filtered Array and transformed Array Elements) mixed in one block of code . It is more readability if we split its responsibilities in lines : 

  $items = $wc_order->get_items()
  $filtered_items = array_filter($items , [call_back_condition])
  $transformed_items = array_map([call_back_transform],$filtered_items)

In this case we looks the condition and we can looks the transform condition in three lines. In the other hand when we hide condition into foreach is more complex for others developers to show where is the position of condition.

Reply:
This would then produce one unnecessary loop over all array elements in the `array_filter` call. The original code is straight-forward, easy to read and does not consume more resources than needed.